[ALLI-7202] [ALLI-7269] [ALLI-7286] Add popover for lateral links in records

EreMaijala opened the pull request:
- Vanha toimimaton auktoriteettien info-laatikko poistettu.
- Kaikki konfiguroitavissa.
- Auktoriteettien tietojen sisällyttäminen popoveriin on mahdollista, mutta on toistaiseksi pois käytöstä. Voidaan ottaa käyttöön, jos koetaan tarpeelliseksi.
- Parissa kohdassa oikaistu toteutuksen pitämiseksi mahdollisimman yksinkertaisena:
  - LIDO-linkitys Kulttuuriympäristön palveluikkunaan on leivottu templaattiin mukaan, eikä sisällä mitään hienoa konfiguraatiota.
  - Finto-rikastus on rakennettu suoraan AJAX-handleriin ilman mitään hienoa plugin-arkkitehtuuria. Voidaan erotella myöhemmin tarpeen mukaan.

Tämä on osittain vähän mutkikas, koska yritin säilyttää yhteensopivuuden vanhojen linkitysten kanssa, eli sama getLinkedField-kutsu näyttää vanhan mallisen linkin, jos popover ei ole käytössä.

EreMaijala committed:
Testing with authority link element.

EreMaijala committed:
Work towards the popover..

EreMaijala committed:
more work

EreMaijala committed:
Try to cope with search type switching...

EreMaijala committed:
Semi-working additional data.

EreMaijala committed:
Improved functionality also when disabled.

EreMaijala committed:
Work on LIDO started.

EreMaijala committed:
Try to make sense of id fields. Add new search templates and AJAX handler.

EreMaijala committed:
More working!

EreMaijala committed:
Merge branch 'dev' into alli-7202-lateral-popover

EreMaijala committed:
qa fixes, partial

EreMaijala committed:
Ontology enrichment etc. fully working, hopefully.

EreMaijala committed:
Update sample config.

EreMaijala committed:
Merge branch 'dev' into alli-7202-lateral-popover

EreMaijala committed:
Omit rest api version from Skosmos base url.

EreMaijala commented:
Käyttöönottoon pitäisi riittää, kun ottaa config.ini.sample-tiedostosta tuon [LinkPopovers]-sektion config.ini-tiedostoon. Linkkivaihtoehtoja voi toki muokata ja lisätä tarpeen mukaan.

EreMaijala commented:
@pasitiis Toiveena käyttöliittymän, toimintojen ja less-muutosten katselmointi. @LuomaJuha toiveena koodin katselmointi mukaanlukien js-palikat.

EreMaijala commented:
Testitietueita:
Kyppi-linkki: museovirasto.00035e8f-1c78-4114-ba20-f9e9a45fd781
Aiheiden ontologiarikastus: 3amk.100435
Tekijä-auktoriteetti: fikka.4347219

EreMaijala committed:
Fix default settings. Add support for multiple IDs.

EreMaijala committed:
Display Kyppi identifiers.

EreMaijala committed:
Update English translations.

LuomaJuha commented on themes/finna2/js/finna-record.js:
```diff
@@ -384,6 +384,89 @@ finna.record = (function finnaRecord() {
+      }
+      infoBounds = container.getBoundingClientRect();
+      if (infoBounds.right > window.innerWidth - 8) {
+        //let marginLeft = -(container.clientWidth - Math.ceil(infoBounds.right - window.innerWidth)) - 8;
```
Pois?

LuomaJuha commented on module/Finna/src/Finna/AjaxHandler/GetFieldInfo.php:
```diff
@@ -0,0 +1,423 @@
+        $params->set('authorityType', $type);
+        $params->set('recordSource', $source);
+        $authority = null;
+        // Uncomment the following if support for displaying authority info is
```
Jääkö tämä koodipohjaan vai pitäisikö olla ennemmin asetus kuin kommentoinnin poisto?

EreMaijala replied:
Lisätty asetus.

LuomaJuha commented on module/Finna/src/Finna/AjaxHandler/GetFieldInfo.php:
```diff
@@ -0,0 +1,423 @@
+    }
+
+    /**
+     * Fetch data for and identifier from Skosmos
```
an?

LuomaJuha commented on module/Finna/src/Finna/View/Helper/Root/Record.php:
```diff
@@ -535,6 +572,39 @@ protected function isAuthorityLinksEnabled()
     }
 
+    /**
+     * Composer additional data string for a link
```
Compose?

LuomaJuha commented on themes/finna2/templates/RecordDriver/DefaultRecord/ajax-field-info.phtml:
```diff
@@ -0,0 +1,93 @@
+          $kyppiLinks[] = [
+            'id' => $matches[2],
+            'type' => $matches[1],
+            'link' => "https://www.kyppi.fi/palveluikkuna/$register/read/asp/r_kohde_det.aspx?KOHDE_ID=" . urlencode($matches[2])
```
Halutaanko kovakoodata linkkiä templaattipohjaan?

EreMaijala replied:
Olisi ollut helppoa, mutta lisäsin nyt asetushässäkän. :)

LuomaJuha commented on themes/finna2/templates/RecordDriver/DefaultRecord/popover-link-element.phtml:
```diff
@@ -0,0 +1,26 @@
+$aAttributes = $this->htmlAttributes(['class' => $this->classes, 'href' => '#']);
+$aAttributes->add('class', 'show-info');
+$aAttributes->add('title', $this->title ?: $this->translate('field_info_link_title'));
+$additionalDataElem = !empty($this->additionalData) ? "<span class=\"field-additional-data\">{$this->additionalDataHtml}</span>" : '';
```
!empty($this->additionalData) pitäisi varmaan tarkistaa additionalDataHtml tässä?

EreMaijala replied:
Korjattu. Tästä seurasi rumuutta, jota vielä täytyy viilata.

LuomaJuha commented on module/Finna/src/Finna/RecordDriver/SolrLido.php:
```diff
@@ -1529,20 +1529,65 @@ public function getSubjectDetails()
+     *
+     * @return array
+     */
+    public function getSubjectPlacesExtended()
```
): array ?

LuomaJuha commented on themes/finna2/templates/RecordDriver/DefaultRecord/ajax-field-info.phtml:
```diff
@@ -0,0 +1,93 @@
+  <div class="additional-content">
+    <?php foreach ($kyppiLinks as $i => $link): ?>
+      <?=($i > 0) ? '<br>' : ''?>
+      <a href="<?=$this->escapeHtmlAttr($link['link'])?>" target="_blank" aria-label="Kulttuuriympäristön palveluikkuna (<?=$this->transEscAttr('external_link')?>, <?=$this->transEscAttr('Open in a new window')?>)">
```
Translaatio Kulttuuriympäristön palveluikkunalle? CULTURAL ENVIRONMENT SERVICE WINDOW | KULTURMILJÖNS TJÄNSTEPORTAL

EreMaijala replied:
Tämä on nyt konfiguraatiossa.

EreMaijala committed:
Fix review issues.

- Makes authority fields configurable.
- Makes external links configurable and moves them to the non-ajax piece.

EreMaijala committed:
Improve screen reader support.

Co-authored-by: Pasi Tiisanoja <pasi.tiisanoja@helsinki.fi>

EreMaijala committed:
Tweak code style.

EreMaijala committed:
Consolidate closing of popover, handle focus better.

EreMaijala committed:
Handle non-authority record id better.

EreMaijala committed:
Set focus only when appropriate.

EreMaijala commented:
Juhan raportoimat ongelmat korjattu, seurauksena uusia asetuksia Kyppi-linkkejä varten. Lisäksi tuotu Pasin tekemät parannukset ruudunlukijoille, siistitty koodia ja laitettu popupin sulkeminen fokusoimaan avauslinkki tarvittaessa.

EreMaijala committed:
Revert some accessibility changes.

EreMaijala committed:
Add source descriptions to places that have an id.

EreMaijala committed:
Focus first link after opening the popover.

LuomaJuha reviewed (approved):
Jätän approvalin, jätin kommentin slackissa yhdestä pienestä löydöksestä mutta omasta mielestä valmis mergettäväksi.

EreMaijala committed:
Add missing template.

EreMaijala committed:
LIDO: Exclude places from subjects to avoid duplicates.

EreMaijala committed:
Use tryMethod to make code checkers happier.

EreMaijala committed:
Fix spacing and extra parenthesis in additional details.

LuomaJuha commented on themes/finna2/templates/RecordDriver/DefaultRecord/link-place.phtml:
```diff
@@ -0,0 +1,3 @@
+<?php // No ending line feed ?>
+<?php $searchRoute = $this->searchAction ?: $this->searchOptions($this->driver->getSearchBackendIdentifier())->getSearchAction(); ?>
+<?=$this->url($searchRoute)?>?lookfor=%22<?=urlencode($this->lookfor)?>%22&amp;type=geographic
```
rivinvaihto

LuomaJuha commented on module/Finna/src/Finna/RecordDriver/SolrLido.php:
```diff
@@ -1526,6 +1526,43 @@ public function getSubjectDetails()
+     *
+     * @return array
+     */
+    public function getAllSubjectHeadingsWithoutPlaces($extended = false)
```
bool $extended? : array?

LuomaJuha reviewed (changes requested):
Pari pientä löytyi.

EreMaijala committed:
Fix review issues.

EreMaijala commented:
Korjattu.

LuomaJuha commented on themes/finna2/templates/RecordDriver/DefaultRecord/link-search-place-id.phtml:
```diff
@@ -0,0 +1,9 @@
+echo $this->url(
+    $this->searchAction ?: $this->searchOptions($this->driver->getSourceIdentifier())->getSearchAction(),
+    [],
+    ['filter[]' => sprintf('%s:("%s")^100000', \Finna\Search\Solr\AuthorityHelper::GEOGRAPHIC_ID_FACET, $id)]
```
Tästä näköjään puuttuu myös 'query' =>

EreMaijala committed:
Fix id search templates.
